docs/naps/7-key-binding-dispatch.md

```diff
@@ -0,0 +1,427 @@
+(nap-7)=
+
+# NAP-7 — Key Binding Dispatch
+
+```{eval-rst}
+:Author: "Kira Evans <mailto:contact@kne42.me>"
+:Created: 2023-07-03
+:Status: Draft
+:Type: Standards Track
+```
+
+## Abstract
+
+With the switching of the internal key binding system to use app-model's representation[^id1], there is discussion as to what exactly constitutes a valid key binding and how conflicts are handled[^id2].
+
+This NAP seeks to clarify and propose a solution for how key bindings will be dispatched according to their priority, enablement, and potential conflicts.
+
+## Motivation and Scope
+
+Plugin developers are able to export commands in their manifest file but cannot similarly set their shortcuts in a code-free way. npe2 provides an option to bind commands to key binding but it is undocumented and unsupported since napari still uses an old dispatch system of chainmaps.
+
+### A more versatile system
+
+The proposed dispatch system would leverage weights to clearly separate default, plugin, and user-defined key binding as well as support a more advanced conditional system which would determine if a keybinding is active/enabled or not. Both of these properties are already part of the specification defined by app-model.
+
+Separation of different sources of key binding makes it easier for a user to determine who set the binding as well as how to restore the default.
+
+Conditional evaluation allows plugin developers to tie their keybinding to a specific viewer state.
+
+### Definitions
+
+**modifier keys** refer to `ctrl`, `shift`, `alt`, and `meta`. `meta` is also known as `cmd`, `win`, or `super` on osx, windows, or linux, respectively
+
+a **base key** is one of the following supported keys:
+- `f1-f12`, `a-z`, `0-9`
+- `` ` ``, `-`, `=`, `[`, `]`, `\`, `;`, `'`, `,`, `.`, `/`
```

Comment:
That is a US centric view. Many of those are characters, not keys, that might be present on the same key. `6` and `-` are the same key on french keyboard for example.

I'm fine having a US centric implementation, but I just want to underline that the nomenclature is important at otherwise it can lead to confusion.

I would suggest pulling a-z0-9 into their own category, and saying something along "keys that support one or more of the following character are base keys".

As far as I understand for the other ones, they may be on the same physical button, (eg, pageupdown, and arrow keys), but needs the fn button button to be pressed, so at keyboard level they register as a different key.

I believe vocabulary and definitions are important otherwise we will talk pass each other, and defining say button/key/characters may help us express things we couldn't do otherwise.

In practice keys/character distinction might be important for the physical layout of  where the shortcuts are. Typically games can use `wasd` for moving character, those are _key_ bindings and not character binding, as on other non-english keyboard, you want the positions of the button to be the same, but the character ends up being `zqsd` on say french keyboard.

Reply:
Oh, that's interesting. I find that explanation confusing as well, so I put down `a **base key** is a key that when pressed without a modifier key, produces one of the following [key codes](https://w3c.github.io/uievents-code/#keyboard-101)`.

Reply:
I think that' better, and thank for your link, it is interesting.

Reply:
I think this excerpt is particularly interesting

<img width="843" alt="Screenshot 2023-07-15 at 2 37 24 PM" src="https://github.com/napari/docs/assets/29165011/44070e9a-df02-49cb-b5e9-e4f9d5a6fc22">

e.g. `KeyCode.KeyQ` produces `'a'` on a French keyboard layout which is not immediately intuitive